packages/go_router/lib/src/shell_state.dart

```diff
@@ -0,0 +1,271 @@
+import 'misc/stateful_navigation_shell.dart';
+
+/// The snapshot of the current state of a [StatefulShellRoute].
+///
+/// Note that this an immutable class, that represents the snapshot of the state
+/// of a StatefulShellRoute at a given point in time. Therefore, instances of
+/// this object should not be stored, but instead fetched fresh when needed,
+/// using the method [StatefulShellRouteState.of].
+@immutable
+class StatefulShellRouteState {
+  /// Constructs a [StatefulShellRouteState].
+  const StatefulShellRouteState({
+    required this.route,
+    required this.branchStates,
+    required this.currentIndex,
+    required void Function(
+            StatefulShellBranchState, UnmodifiableRouteMatchList?)
+        switchActiveBranch,
+    required void Function(StatefulShellBranchState? branchState,
+            bool navigateToDefaultLocation)
+        resetState,
+  })  : _switchActiveBranch = switchActiveBranch,
+        _resetState = resetState;
+
+  /// Constructs a copy of this [StatefulShellRouteState], with updated values
+  /// for some of the fields.
+  StatefulShellRouteState copy(
+      {List<StatefulShellBranchState>? branchStates, int? currentIndex}) {
+    return StatefulShellRouteState(
+      route: route,
+      branchStates: branchStates ?? this.branchStates,
+      currentIndex: currentIndex ?? this.currentIndex,
+      switchActiveBranch: _switchActiveBranch,
+      resetState: _resetState,
+    );
+  }
+
+  /// The associated [StatefulShellRoute]
+  final StatefulShellRoute route;
+
+  /// The state for all separate route branches associated with a
+  /// [StatefulShellRoute].
+  final List<StatefulShellBranchState> branchStates;
+
+  /// The state associated with the current [StatefulShellBranch].
+  StatefulShellBranchState get currentBranchState => branchStates[currentIndex];
+
+  /// The index of the currently active [StatefulShellBranch].
+  ///
+  /// Corresponds to the index of the branch in the List returned from
+  /// branchBuilder of [StatefulShellRoute].
+  final int currentIndex;
+
+  /// The Navigator key of the current navigator.
+  GlobalKey<NavigatorState> get currentNavigatorKey =>
+      currentBranchState.branch.navigatorKey;
+
+  final void Function(StatefulShellBranchState, UnmodifiableRouteMatchList?)
+      _switchActiveBranch;
+
+  final void Function(
+          StatefulShellBranchState? branchState, bool navigateToDefaultLocation)
+      _resetState;
+
+  /// Gets the [Widget]s representing each of the shell branches.
+  ///
+  /// The Widget returned from this method contains the [Navigator]s of the
+  /// branches. Note that the Widgets returned by this method should only be
+  /// added to the widget tree if using a custom branch container Widget
+  /// implementation, where the child parameter in the [ShellRouteBuilder] of
+  /// the [StatefulShellRoute] is ignored (i.e. not added to the widget tree).
+  /// See [StatefulShellBranchState.child].
+  List<Widget> get children =>
```

Comment:
This is a weird API, I don't think we should encourage this. only one child should be on the screen at one time. Also what would the children be if the branch does not preload?

Reply:
I think this is partially covered by above answers, and I agree the field should probably be moved (and it has been in the other branch). However the fact that multiple children (i.e. Navigators) are on the screen (Widget tree) at one time is key to making this whole thing work, otherwise we wouldn't be able to maintain state in the Navigators. Of course, only one child will be _visible_ at a time, but the other children must still be in the `Widget` tree, albeit `Offstage`.